sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,178 @@
+
+/**
+ * This flag is used to verify that the instruction access fault has occurred,
+ * and has been successfully serviced. Declared as volatile, because it is
+ * referenced in the fault handler, as well as the main test flow.
+ */
+static volatile bool instruction_access_fault;
+
+static const uint32_t kRetRamStartAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR;
+static const uint32_t kRetRamEndAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR +
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_SIZE_BYTES - 1;
+
+/**
+ * Buffer filled with `0` (valid `unimp`) instructions.
+ *
+ * When Retention SRAM is disabled, executing this buffer will result in an
+ * Instruction Access Fault, which constitutes a successful test. Otherwise,
+ * it will result in an Illegal Instruction Fault, which constitutes a failure.
+ */
+static const uint32_t execution_test_instructions[2];
+
+/**
+ * Determines and sets the return address from an exception handler.
+ *
+ * The exception cause instruction, could be either standard or compressed,
+ * which in turn means that the return address (next instruction after fault)
+ * will be either at the 2byte (16bit) or 4byte (32bit) offset.
+ *
+ * @param indirection `NULL` to calculate return address relative to the
+ *                    exception cause instruction (default), or an alternative
+ *                    return address.
+ */
+static void fault_handler_set_return_addr(uintptr_t *indirection) {
+  uint32_t mepc = ibex_mepc_read();
+  LOG_INFO("Fault address: mepc = 0x%x", mepc);
+
+  if (indirection == NULL) {
+    // Check if the two least significant bits of the instruction are b11 (0x3),
+    // which means that the trapped instruction is not compressed
+    // (32bits = 4bytes), otherwise (16bits = 2bytes).
+    //
+    // NOTE:
+    // with RISC-V "c" (compressed instructions extension), 32bit
+    // instructions can start on 16bit boundary.
+    //
+    // Please see:
+    // "“C” Standard Extension for Compressed Instructions, Version 2.0",
+    // section 16.1.
+    bool not_compressed = (mepc & 0x3) == 0x3;
+    mepc = not_compressed ? (mepc + 4) : (mepc + 2);
+  } else {
+    mepc = *indirection;
+    LOG_INFO("Indirect return (one up from the next stack frame)");
+  }
+
+  LOG_INFO("Return address: mepc = 0x%x", mepc);
+  ibex_mepc_write(mepc);
+}
+
+/**
+ * Handles an exception.
+ *
+ * This exception handler keeps default implementations for faults that are not
+ * relevant in terms of the test. The only two fault handlers that are
+ * overridden, are the Instruction Access Fault and Illegal Instruction Fault.
+ *
+ * The controlled fault originates in the retention SRAM, which means that
+ * normally the return address would be calculated relevant to the offending
+ * instruction. However, due to execution from retention SRAM being permanently
+ * disabled, this approach would not work.
+ * The return address contained in this stack frame. In other words
```

Comment:
`s/The return address contained in this stack frame./Instead control needs to be returned to the caller./`